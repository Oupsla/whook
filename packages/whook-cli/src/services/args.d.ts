import { LogService } from 'common-services';
import { WhookArgsTypes } from '..';
declare const _default: typeof initArgs;
export default _default;
interface WhookCommandNamedArgs {
  [name: string]: WhookArgsTypes;
}
interface WhookCommandArgsRest {
  _?: string[];
}
export declare type WhookCommandArgs = WhookCommandNamedArgs &
  WhookCommandArgsRest;
declare function initArgs({
  ARGS,
  log,
}: {
  ARGS: string[];
  log: LogService;
}): Promise<WhookCommandArgs>;
