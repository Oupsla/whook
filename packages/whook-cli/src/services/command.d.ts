declare const _default: typeof initCommand;
export default _default;
declare function initCommand({
  commandHandler,
  log,
}: {
  commandHandler: any;
  log: any;
}): Promise<() => Promise<void>>;
