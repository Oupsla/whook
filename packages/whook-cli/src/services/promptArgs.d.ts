import _inquirer from 'inquirer';
import { LogService } from 'common-services';
import { WhookCommandArgs } from './args';
declare const _default: typeof initPromptArgs;
export default _default;
export declare type WhookArgsTypes = string | boolean | number;
export declare type WhookCommandHandler = () => Promise<void>;
export declare type WhookCommandDefinitionArguments = {
  type: 'object';
  required?: string[];
  additionalProperties: false;
  properties: {
    [name: string]: {
      type: 'boolean' | 'string' | 'number';
      description: string;
      enum?: WhookArgsTypes[];
      default?: string | boolean | number;
    };
  };
};
export declare type WhookCommandDefinition = {
  description: string;
  example: string;
  arguments: WhookCommandDefinitionArguments;
};
export declare type PromptArgs = () => Promise<{
  [name: string]: any;
}>;
declare function initPromptArgs({
  COMMAND_DEFINITION,
  args,
  inquirer,
  log,
}: {
  COMMAND_DEFINITION: WhookCommandDefinition;
  args: WhookCommandArgs;
  inquirer?: typeof _inquirer;
  log?: LogService;
}): Promise<PromptArgs>;
