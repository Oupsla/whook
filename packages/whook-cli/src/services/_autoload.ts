import path from 'path';
import initAutoloader from '@whook/whook/dist/services/_autoload';
import {
  wrapInitializer,
  alsoInject,
  service,
  constant,
  name,
} from 'knifecycle';
import YError from 'yerror';

// Needed to avoid messing up babel builds ðŸ¤·
const _require = require;

export default alsoInject(
  ['args', 'PROJECT_SRC', 'WHOOK_PLUGINS_PATHS', 'log'],
  wrapInitializer(
    async (
      { PROJECT_SRC, WHOOK_PLUGINS_PATHS, args, log, require = _require },
      $autoload,
    ) => {
      log('debug', 'ðŸ¤– - Wrapping the whook autoloader.');

      const commandName = args._[0];
      let commandModule;

      if (!commandName) {
        commandModule = {
          default: service(
            async () => async () => {
              log('warning', `No command given in argument.`);
            },
            'commandHandler',
          ),
          definition: {},
        };
      } else {
        [PROJECT_SRC, ...WHOOK_PLUGINS_PATHS].some(basePath => {
          const finalPath = path.join(basePath, 'commands', commandName);
          try {
            commandModule = require(finalPath);
            return true;
          } catch (err) {
            log('debug', `Command "${commandName}" not found in: ${finalPath}`);
            log('debug-stack', err.stack);
          }
        });
      }

      if (!commandModule) {
        commandModule = {
          default: service(
            async () => async () => {
              log('warning', `Command "${commandName}" not found.`);
            },
            'commandHandler',
          ),
          definition: {},
        };
      }

      if (!commandModule.default) {
        throw new YError('E_NO_COMMAND_HANDLER', commandName);
      }
      if (!commandModule.definition) {
        throw new YError('E_NO_COMMAND_DEFINITION', commandName);
      }

      return async serviceName => {
        if (serviceName === 'COMMAND_DEFINITION') {
          return {
            initializer: constant(serviceName, commandModule.definition),
            path: `definition://${commandName}`,
          };
        }
        if (serviceName === 'commandHandler') {
          return {
            initializer: name('commandHandler', commandModule.default),
            path: `command://${commandName}`,
          };
        }

        return $autoload(serviceName);
      };
    },
    initAutoloader,
  ),
);
