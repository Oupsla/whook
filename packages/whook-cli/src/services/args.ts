import { autoService, options } from 'knifecycle';
import parseArgs from 'yargs-parser';
import { LogService } from 'common-services';
import { WhookArgsTypes } from '..';

export default options({ singleton: true }, autoService(initArgs));

interface WhookCommandNamedArgs {
  [name: string]: WhookArgsTypes;
}
interface WhookCommandArgsRest {
  _?: string[];
}
export type WhookCommandArgs = WhookCommandNamedArgs & WhookCommandArgsRest;

async function initArgs({
  ARGS,
  log,
}: {
  ARGS: string[];
  log: LogService;
}): Promise<WhookCommandArgs> {
  const args = parseArgs(ARGS.slice(2));

  log('debug', 'ðŸ›  - Parsed args:', args);
  return args;
}
