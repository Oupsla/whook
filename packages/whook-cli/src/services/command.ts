import { autoService, options } from 'knifecycle';

export default options({ singleton: true }, autoService(initCommand));

async function initCommand({ commandHandler, log }) {
  return async function commandRunner() {
    try {
      await commandHandler();
    } catch (err) {
      if (err.code === 'E_BAD_ARGS') {
        log('stack', err.stack);
        if (err.params[0][0].keyword === 'required') {
          if (err.params[0][0].params.missingProperty) {
            log(
              'error',
              `Argument "${err.params[0][0].params.missingProperty}" is required.`,
            );
            return;
          }
        }
        if (err.params[0][0].keyword === 'additionalProperties') {
          if (err.params[0][0].params.additionalProperty === '_') {
            log('error', 'No anonymous arguments allowed.');
            return;
          }
          if (err.params[0][0].params.additionalProperty) {
            log(
              'error',
              `Argument "${err.params[0][0].params.additionalProperty}" not allowed.`,
            );
            return;
          }
        }
        log(
          'error',
          'Error parsing arguments: ',
          err.params[0][0].message,
          err.params[0][0].params,
        );
        return;
      }
      throw err;
    }
  };
}
