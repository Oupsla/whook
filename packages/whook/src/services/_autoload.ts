import { identity, noop, compose } from '../libs/utils';
import {
  initializer,
  constant,
  name,
  Injector,
  Autoloader,
  ServiceInitializer,
} from 'knifecycle';
import {
  flattenOpenAPI,
  getOpenAPIOperations,
} from '@whook/http-router/dist/utils';
import path from 'path';
import YError from 'yerror';
import { CONFIGSService } from './CONFIGS';
import { LogService } from 'common-services';
import { WhookHandler } from '@whook/http-transaction';

export const HANDLER_REG_EXP = /^(head|get|put|post|delete|options|handle)[A-Z][a-zA-Z0-9]+/;

// Needed to avoid messing up babel builds ðŸ¤·
const _require = require;
const _resolve = require.resolve;

export interface WhookWrapper<D, S extends WhookHandler> {
  (initializer: ServiceInitializer<D, S>): ServiceInitializer<D, S>;
}
/* Architecture Note #5.7.1: WhookServiceMap
Whook exports a `WhookServiceMap` type to help you ensure yours are valid.
*/
export type WhookServiceMap = { [name: string]: string };
/* Architecture Note #5.6.1: WhookInitializerMap
Whook exports a `WhookInitializerMap` type to help you ensure yours are valid.
*/
export type WhookInitializerMap = { [name: string]: string };

/* Architecture Note #5: `$autoload` service
The default Whook autoloader provides a simple way to
 load the constants, services and handlers of a Whook
 project automatically from the installed whook plugins.
*/
export default initializer(
  {
    name: '$autoload',
    type: 'service',
    inject: [
      'PROJECT_SRC',
      'WHOOK_PLUGINS_PATHS',
      '$injector',
      '?SERVICE_NAME_MAP',
      '?INITIALIZER_PATH_MAP',
      '?WRAPPERS',
      '?CONFIGS',
      '?log',
    ],
    options: { singleton: true },
  },
  initAutoload,
);

/**
 * Initialize the Whook default DI autoloader
 * @param  {Object}   services
 * The services `$autoload` depends on
 * @param  {Object}   services.PROJECT_SRC
 * The project source directory
 * @param  {Object}   services.WHOOK_PLUGINS
 * The plugins to load services from
 * @param  {Object}   services.$injector
 * An injector for internal dynamic services loading
 * @param  {Object}   [services.SERVICE_NAME_MAP={}]
 * An optional object to map services names to other names
 * @param  {Object}   [services.INITIALIZER_PATH_MAP={}]
 * An optional object to map non-autoloadable initializers
 * @param  {Array}   [services.WRAPPERS]
 * An optional list of wrappers to inject
 * @param  {Array}   [services.CONFIGS]
 * Optional CONFIGS object to inject
 * @param  {Object}   [log=noop]
 * An optional logging service
 * @return {Promise<Function>}
 * A promise of the autoload function.
 */
async function initAutoload<D, S>({
  PROJECT_SRC,
  WHOOK_PLUGINS_PATHS = [],
  $injector,
  SERVICE_NAME_MAP,
  INITIALIZER_PATH_MAP = {},
  WRAPPERS,
  CONFIGS,
  log = noop,
  require = _require,
  resolve = _resolve,
}: {
  PROJECT_SRC: string;
  WHOOK_PLUGINS_PATHS?: string[];
  $injector: Injector<any>;
  SERVICE_NAME_MAP?: WhookServiceMap;
  INITIALIZER_PATH_MAP?: WhookInitializerMap;
  WRAPPERS?: WhookWrapper<D, WhookHandler>[];
  CONFIGS?: CONFIGSService;
  log?: LogService;
  require?: typeof _require;
  resolve?: typeof _resolve;
}): Promise<Autoloader<any>> {
  log('debug', 'ðŸ¤– - Initializing the `$autoload` service.');

  /* Architecture Note #5.3: API auto loading
  We cannot inject the `API` in the auto loader since
   it is dynamically loaded so doing this during the auto
   loader initialization.
  */
  let API;
  const getAPI = (() => {
    return async () => {
      if (!API) {
        // eslint-disable-next-line
        API = (await $injector(['API'])).API;
      }
      return API;
    };
  })();

  /* Architecture Note #5.2: Wrappers auto loading support
  We cannot inject the `WRAPPERS` in the auto loader when
   it is dynamically loaded so giving a second chance here
   for `WRAPPERS` to be set.
  */
  const doWrapHandler = (WRAPPERS => {
    let wrapHandler;

    return async handlerInitializer => {
      if (!WRAPPERS) {
        // eslint-disable-next-line
        WRAPPERS = (await $injector(['WRAPPERS'])).WRAPPERS || [];
      }
      wrapHandler =
        wrapHandler || (WRAPPERS.length ? compose(...WRAPPERS) : identity);

      return wrapHandler(handlerInitializer);
    };
  })(WRAPPERS);

  const mapInjectedName = async injectedName => {
    if ('SERVICE_NAME_MAP' === injectedName) {
      return injectedName;
    }
    if (!SERVICE_NAME_MAP) {
      // eslint-disable-next-line
      SERVICE_NAME_MAP = (await $injector(['SERVICE_NAME_MAP']))
        .SERVICE_NAME_MAP;
    }

    return SERVICE_NAME_MAP[injectedName] || injectedName;
  };

  return $autoload;

  /**
   * Autoload an initializer from its name
   * @param  {String}  The dependency name
   * @return {Promise<Object>}
   * A promise resolving whith the actual autoloader definition.
   *  An Object containing the `path`, `name` and the `initializer`
   *  in its properties.
   */
  async function $autoload(injectedName) {
    /* Architecture Note #5.7: Service name mapping
    In order to be able to substituate easily a service per another
     one can specify a mapping between a service and its substitution.
    */
    const resolvedName = await mapInjectedName(injectedName);

    if (resolvedName !== injectedName) {
      log(
        'debug',
        `ðŸ“– - Using SERVICE_NAME_MAP to route "${injectedName}" to "${resolvedName}".`,
      );
    }

    /* Architecture Note #5.1: Configuration auto loading
    Loading the configuration files is done according to the `NODE_ENV`
     environment variable. It basically requires a configuration hash
     where the keys are Knifecycle constants.

    Let's load the configuration files as a convenient way
     to create constants on the fly
    */
    if ('CONFIGS' !== resolvedName) {
      if (!CONFIGS) {
        // eslint-disable-next-line
        CONFIGS = (await $injector(['CONFIGS'])).CONFIGS;
      }
    }

    /* Architecture Note #5.4: Constants
    First of all the autoloader looks for constants in the
     previously loaded `CONFIGS` configurations hash.
    */
    if (CONFIGS && CONFIGS[resolvedName]) {
      return {
        name: resolvedName,
        path: 'internal://' + resolvedName,
        initializer: constant(resolvedName, CONFIGS[resolvedName]),
      };
    }

    const isHandler = HANDLER_REG_EXP.test(resolvedName);
    const isWrappedHandler = resolvedName.endsWith('Wrapped');

    /* Architecture Note #5.5: Handlers map
    Here, we build the handlers map needed by the router by injecting every
     handler required by the API.
    */
    if ('HANDLERS' === resolvedName) {
      const handlerNames = [
        ...new Set(
          (
            await getOpenAPIOperations(await flattenOpenAPI(await getAPI()))
          ).map(operation => operation.operationId),
        ),
      ].map(handlerName => `${handlerName}>${handlerName}Wrapped`);

      return {
        name: resolvedName,
        initializer: initializer(
          {
            name: 'HANDLERS',
            inject: handlerNames,
            type: 'service',
            options: {
              singleton: true,
            },
          },
          async HANDLERS => HANDLERS,
        ),
        path: 'internal://' + resolvedName,
      };
    }

    /* Architecture Note #5.6: Service/handler loading
    Finally, we either require the handler/service module if
     none of the previous strategies applyed.
    */
    const modulePath =
      INITIALIZER_PATH_MAP[resolvedName] ||
      [PROJECT_SRC, ...WHOOK_PLUGINS_PATHS].reduce(
        (finalModulePath, basePath) => {
          if (finalModulePath) {
            return finalModulePath;
          }

          const finalPath = path.join(
            basePath,
            isHandler ? 'handlers' : 'services',
            isWrappedHandler
              ? resolvedName.replace(/Wrapped$/, '')
              : resolvedName,
          );

          try {
            return resolve(finalPath);
          } catch (err) {
            log(
              'debug',
              `ðŸš« - Service "${resolvedName}" not found in: ${finalPath}`,
            );
            return null;
          }
        },
        null,
      );

    /* Architecture Note #5.8: Initializer path mapping
    In order to be able to load a service from a given path map
     one can directly specify a path to use for its resolution.
    */
    if (INITIALIZER_PATH_MAP[resolvedName]) {
      log(
        'debug',
        `ðŸ“– - Using INITIALIZER_PATH_MAP to resolve the "${resolvedName}" module path.`,
      );
    }

    if (!modulePath) {
      throw new YError('E_UNMATCHED_DEPENDENCY', resolvedName);
    }

    log('debug', `ðŸ’¿ - Service "${resolvedName}" found in: ${modulePath}`);

    const resolvedInitializer = await require(modulePath).default;

    log(
      'debug',
      `ðŸ’¿ - Loading ${injectedName} initializer${
        resolvedName !== injectedName ? ` via ${resolvedName} resolution` : ''
      } from ${modulePath}.`,
    );

    return {
      name: resolvedName,
      path: modulePath,
      initializer: isWrappedHandler
        ? name(resolvedName, await doWrapHandler(resolvedInitializer))
        : injectedName !== resolvedName
        ? name(injectedName, resolvedInitializer)
        : resolvedInitializer,
    };
  }
}
