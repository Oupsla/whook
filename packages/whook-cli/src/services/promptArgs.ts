import parseArgs from 'yargs-parser';
import { autoService, options } from 'knifecycle';
import _inquirer from 'inquirer';
import { LogService } from 'common-services';
import { noop } from '@whook/whook';
import { WhookCommandArgs } from './args';

export default options({ singleton: true }, autoService(initPromptArgs));

export type WhookArgsTypes = string | boolean | number;
export type WhookCommandHandler = () => Promise<void>;
// Subset of JSON Schema types so not using @types/json-schema
export type WhookCommandDefinitionArguments = {
  type: 'object';
  required?: string[];
  additionalProperties: false;
  properties: {
    [name: string]: {
      type: 'boolean' | 'string' | 'number';
      description: string;
      enum?: WhookArgsTypes[];
      default?: string | boolean | number;
    };
  };
};
export type WhookCommandDefinition = {
  description: string;
  example: string;
  arguments: WhookCommandDefinitionArguments;
};
export type PromptArgs = () => Promise<{ [name: string]: any }>;

async function initPromptArgs({
  COMMAND_DEFINITION,
  args,
  inquirer = _inquirer,
  log = noop,
}: {
  COMMAND_DEFINITION: WhookCommandDefinition;
  args: WhookCommandArgs;
  inquirer?: typeof _inquirer;
  log?: LogService;
}): Promise<PromptArgs> {
  log('debug', 'ðŸ›  - Initializing promptArgs service');

  return async () => {
    const questions = (COMMAND_DEFINITION.arguments.required || []).reduce(
      (questions, propertyName) => {
        if ('undefined' === typeof args[propertyName]) {
          const propertyDefinition =
            COMMAND_DEFINITION.arguments.properties[propertyName];

          if (propertyDefinition.type === 'boolean') {
            return [
              ...questions,
              {
                name: propertyName,
                type: 'checkbox',
                message: propertyDefinition.description || '',
              },
            ];
          }
          if (['string', 'number'].includes(propertyDefinition.type)) {
            if (propertyDefinition.enum) {
              return [
                ...questions,
                {
                  name: propertyName,
                  type: 'list',
                  message: propertyDefinition.description || '',
                  choices: propertyDefinition.enum,
                },
              ];
            }

            return [
              ...questions,
              {
                name: propertyName,
                type: 'input',
                message: propertyDefinition.description || '',
              },
            ];
          }
        }
        return questions;
      },
      [],
    );

    if (0 === questions.length) {
      return args;
    }

    return {
      ...args,
      ...(await inquirer.prompt(questions)),
    };
  };
}
