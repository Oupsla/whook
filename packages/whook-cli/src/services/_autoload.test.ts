import { resolveWhookPlugins } from '@whook/whook/dist/services/_autoload';
import YError from 'yerror';

describe('$autoload', () => {
  const log = jest.fn();
  const requireMock = jest.fn();
  const $baseAutoload = jest.fn();

  beforeEach(() => {
    jest.resetModules();
    log.mockReset();
    requireMock.mockReset();
    $baseAutoload.mockReset();
  });

  it('should warn with no command name', async () => {
    jest.doMock('@whook/whook/dist/services/_autoload', () => {
      const _default = async () => $baseAutoload;

      _default.default = _default;
      _default.resolveWhookPlugins = resolveWhookPlugins;
      return _default;
    });

    const initAutoload = require('./_autoload').default;
    const $autoload = await initAutoload({
      PROJECT_SRC: '/home/whoiam/projects/my-whook-project/src',
      WHOOK_PLUGINS_PATHS: ['/var/lib/node/node_modules/@whook/whook/src'],
      args: { _: [] },
      require: requireMock,
      log,
    });
    const { path, initializer } = await $autoload('commandHandler');
    const command = await initializer();
    const result = await command();

    expect({
      path,
      result,
      requireCalls: requireMock.mock.calls,
      baseAutoloaderCalls: $baseAutoload.mock.calls,
      logCalls: log.mock.calls.filter(args => 'debug-stack' !== args[0]),
    }).toMatchSnapshot();
  });

  it('should warn with not found commands', async () => {
    jest.doMock('@whook/whook/dist/services/_autoload', () => {
      const _default = async () => $baseAutoload;

      _default.default = _default;
      _default.resolveWhookPlugins = resolveWhookPlugins;
      return _default;
    });
    requireMock.mockImplementationOnce(() => {
      throw new YError('E_NO_MODULE');
    });

    const initAutoload = require('./_autoload').default;
    const $autoload = await initAutoload({
      PROJECT_SRC: '/home/whoiam/projects/my-whook-project/src',
      WHOOK_PLUGINS_PATHS: ['/var/lib/node/node_modules/@whook/whook/src'],
      args: { _: ['myCommand'] },
      require: requireMock,
      log,
    });
    const { path, initializer } = await $autoload('commandHandler');
    const command = await initializer();
    const result = await command();

    expect({
      path,
      result,
      requireCalls: requireMock.mock.calls,
      baseAutoloaderCalls: $baseAutoload.mock.calls,
      logCalls: log.mock.calls.filter(args => 'debug-stack' !== args[0]),
    }).toMatchSnapshot();
  });

  it('should work with project commands', async () => {
    jest.doMock('@whook/whook/dist/services/_autoload', () => {
      const _default = async () => $baseAutoload;

      _default.default = _default;
      _default.resolveWhookPlugins = resolveWhookPlugins;
      return _default;
    });
    requireMock.mockReturnValueOnce({
      default: async () => async () => log('warning', 'Command called!'),
      definition: {},
    });

    const initAutoload = require('./_autoload').default;
    const $autoload = await initAutoload({
      PROJECT_SRC: '/home/whoiam/projects/my-whook-project/src',
      WHOOK_PLUGINS_PATHS: ['/var/lib/node/node_modules/@whook/whook/src'],
      args: { _: ['myCommand'] },
      require: requireMock,
      log,
    });
    const { path, initializer } = await $autoload('commandHandler');
    const command = await initializer();
    const result = await command();

    expect({
      path,
      result,
      requireCalls: requireMock.mock.calls,
      baseAutoloaderCalls: $baseAutoload.mock.calls,
      logCalls: log.mock.calls.filter(args => 'debug-stack' !== args[0]),
    }).toMatchSnapshot();
  });

  it('should work with whook-cli commands', async () => {
    jest.doMock('@whook/whook/dist/services/_autoload', () => {
      const _default = async () => $baseAutoload;

      _default.default = _default;
      _default.resolveWhookPlugins = resolveWhookPlugins;
      return _default;
    });
    requireMock.mockImplementationOnce(() => {
      throw new Error('ENOENT');
    });
    requireMock.mockImplementationOnce(() => ({
      default: async () => async () => log('warning', 'Command called!'),
      definition: {},
    }));

    const initAutoload = require('./_autoload').default;
    const $autoload = await initAutoload({
      PROJECT_SRC: '/home/whoiam/projects/my-whook-project/src',
      WHOOK_PLUGINS_PATHS: ['/var/lib/node/node_modules/@whook/whook/src'],
      args: { _: ['myCommand'] },
      require: requireMock,
      log,
    });
    const { path, initializer } = await $autoload('commandHandler');
    const command = await initializer();
    const result = await command();

    expect({
      path,
      result,
      requireCalls: requireMock.mock.calls,
      baseAutoloaderCalls: $baseAutoload.mock.calls,
      logCalls: log.mock.calls.filter(args => 'debug-stack' !== args[0]),
    }).toMatchSnapshot();
  });

  it('should work with bad commands', async () => {
    jest.doMock('@whook/whook/dist/services/_autoload', () => {
      const _default = async () => $baseAutoload;

      _default.default = _default;
      _default.resolveWhookPlugins = resolveWhookPlugins;
      return _default;
    });
    requireMock.mockImplementationOnce(() => {
      throw new Error('ENOENT');
    });
    requireMock.mockImplementationOnce(() => {
      throw new Error('ENOENT');
    });
    $baseAutoload.mockResolvedValueOnce({
      initializer: async () => async () => {},
      path: 'mocked://service',
    });

    const initAutoload = require('./_autoload').default;
    const $autoload = await initAutoload({
      PROJECT_SRC: '/home/whoiam/projects/my-whook-project/src',
      WHOOK_PLUGINS_PATHS: ['/var/lib/node/node_modules/@whook/whook/src'],
      args: { _: ['myCommand'] },
      require: requireMock,
      log,
    });
    const { path, initializer } = await $autoload('commandHandler');
    const command = await initializer();
    const result = await command();

    expect({
      path,
      result,
      requireCalls: requireMock.mock.calls,
      baseAutoloaderCalls: $baseAutoload.mock.calls,
      logCalls: log.mock.calls.filter(args => 'debug-stack' !== args[0]),
    }).toMatchSnapshot();
  });

  it('should delegate to whook $autoloader', async () => {
    jest.doMock('@whook/whook/dist/services/_autoload', () => {
      const _default = async () => $baseAutoload;

      _default.default = _default;
      _default.resolveWhookPlugins = resolveWhookPlugins;
      return _default;
    });
    requireMock.mockImplementationOnce(() => {
      throw new Error('ENOENT');
    });
    requireMock.mockImplementationOnce(() => {
      throw new Error('ENOENT');
    });
    $baseAutoload.mockResolvedValueOnce({
      initializer: async () => async () => {},
      path: 'mocked://service',
    });

    const initAutoload = require('./_autoload').default;
    const $autoload = await initAutoload({
      PROJECT_SRC: '/home/whoiam/projects/my-whook-project/src',
      WHOOK_PLUGINS_PATHS: ['/var/lib/node/node_modules/@whook/whook/src'],
      args: { _: ['myCommand'] },
      require: requireMock,
      log,
    });
    const { path, initializer } = await $autoload('anotherService');
    const command = await initializer();
    const result = await command();

    expect({
      path,
      result,
      requireCalls: requireMock.mock.calls,
      baseAutoloaderCalls: $baseAutoload.mock.calls,
      logCalls: log.mock.calls.filter(args => 'debug-stack' !== args[0]),
    }).toMatchSnapshot();
  });

  describe('should fail', () => {
    it('with no command handler', async () => {
      jest.doMock('@whook/whook/dist/services/_autoload', () => {
        const _default = async () => $baseAutoload;

        _default.default = _default;
        _default.resolveWhookPlugins = resolveWhookPlugins;
        return _default;
      });
      requireMock.mockReturnValueOnce({});

      const initAutoload = require('./_autoload').default;

      try {
        await initAutoload({
          PROJECT_SRC: '/home/whoiam/projects/my-whook-project/src',
          WHOOK_PLUGINS_PATHS: ['/var/lib/node/node_modules/@whook/whook/src'],
          args: { _: ['myCommand'] },
          require: requireMock,
          log,
        });
        throw new YError('E_UNEXPECTED_SUCCESS');
      } catch (err) {
        expect({
          errorCode: err.code,
          errorParams: err.params,
          requireCalls: requireMock.mock.calls,
          baseAutoloaderCalls: $baseAutoload.mock.calls,
          logCalls: log.mock.calls.filter(args => 'debug-stack' !== args[0]),
        }).toMatchSnapshot();
      }
    });

    it('with no command definition', async () => {
      jest.doMock('@whook/whook/dist/services/_autoload', () => {
        const _default = async () => $baseAutoload;

        _default.default = _default;
        _default.resolveWhookPlugins = resolveWhookPlugins;
        return _default;
      });
      requireMock.mockReturnValueOnce({
        default: async () => {},
      });

      const initAutoload = require('./_autoload').default;

      try {
        await initAutoload({
          PROJECT_SRC: '/home/whoiam/projects/my-whook-project/src',
          WHOOK_PLUGINS_PATHS: ['/var/lib/node/node_modules/@whook/whook/src'],
          args: { _: ['myCommand'] },
          require: requireMock,
          log,
        });
        throw new YError('E_UNEXPECTED_SUCCESS');
      } catch (err) {
        expect({
          errorCode: err.code,
          errorParams: err.params,
          requireCalls: requireMock.mock.calls,
          baseAutoloaderCalls: $baseAutoload.mock.calls,
          logCalls: log.mock.calls.filter(args => 'debug-stack' !== args[0]),
        }).toMatchSnapshot();
      }
    });
  });
});
