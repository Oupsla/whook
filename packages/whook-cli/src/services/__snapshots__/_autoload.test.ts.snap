// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`$autoload should delegate to whook $autoloader 1`] = `
Object {
  "baseAutoloaderCalls": Array [
    Array [
      "anotherService",
    ],
  ],
  "logCalls": Array [
    Array [
      "debug",
      "🤖 - Wrapping the whook autoloader.",
    ],
    Array [
      "debug",
      "Command \\"myCommand\\" not found in: /home/whoiam/projects/my-whook-project/src/commands/myCommand",
    ],
    Array [
      "debug",
      "Command \\"myCommand\\" not found in: /var/lib/node/node_modules/@whook/whook/src/commands/myCommand",
    ],
  ],
  "path": "mocked://service",
  "requireCalls": Array [
    Array [
      "/home/whoiam/projects/my-whook-project/src/commands/myCommand",
    ],
    Array [
      "/var/lib/node/node_modules/@whook/whook/src/commands/myCommand",
    ],
  ],
  "result": undefined,
}
`;

exports[`$autoload should fail with no command definition 1`] = `
Object {
  "baseAutoloaderCalls": Array [],
  "errorCode": "E_NO_COMMAND_DEFINITION",
  "errorParams": Array [
    "myCommand",
  ],
  "logCalls": Array [
    Array [
      "debug",
      "🤖 - Wrapping the whook autoloader.",
    ],
  ],
  "requireCalls": Array [
    Array [
      "/home/whoiam/projects/my-whook-project/src/commands/myCommand",
    ],
  ],
}
`;

exports[`$autoload should fail with no command handler 1`] = `
Object {
  "baseAutoloaderCalls": Array [],
  "errorCode": "E_NO_COMMAND_HANDLER",
  "errorParams": Array [
    "myCommand",
  ],
  "logCalls": Array [
    Array [
      "debug",
      "🤖 - Wrapping the whook autoloader.",
    ],
  ],
  "requireCalls": Array [
    Array [
      "/home/whoiam/projects/my-whook-project/src/commands/myCommand",
    ],
  ],
}
`;

exports[`$autoload should warn with no command name 1`] = `
Object {
  "baseAutoloaderCalls": Array [],
  "logCalls": Array [
    Array [
      "debug",
      "🤖 - Wrapping the whook autoloader.",
    ],
    Array [
      "warning",
      "No command given in argument.",
    ],
  ],
  "path": "command://undefined",
  "requireCalls": Array [],
  "result": undefined,
}
`;

exports[`$autoload should warn with not found commands 1`] = `
Object {
  "baseAutoloaderCalls": Array [],
  "logCalls": Array [
    Array [
      "debug",
      "🤖 - Wrapping the whook autoloader.",
    ],
    Array [
      "debug",
      "Command \\"myCommand\\" not found in: /home/whoiam/projects/my-whook-project/src/commands/myCommand",
    ],
    Array [
      "warning",
      "Command \\"myCommand\\" not found.",
    ],
  ],
  "path": "command://myCommand",
  "requireCalls": Array [
    Array [
      "/home/whoiam/projects/my-whook-project/src/commands/myCommand",
    ],
    Array [
      "/var/lib/node/node_modules/@whook/whook/src/commands/myCommand",
    ],
  ],
  "result": undefined,
}
`;

exports[`$autoload should work with bad commands 1`] = `
Object {
  "baseAutoloaderCalls": Array [],
  "logCalls": Array [
    Array [
      "debug",
      "🤖 - Wrapping the whook autoloader.",
    ],
    Array [
      "debug",
      "Command \\"myCommand\\" not found in: /home/whoiam/projects/my-whook-project/src/commands/myCommand",
    ],
    Array [
      "debug",
      "Command \\"myCommand\\" not found in: /var/lib/node/node_modules/@whook/whook/src/commands/myCommand",
    ],
    Array [
      "warning",
      "Command \\"myCommand\\" not found.",
    ],
  ],
  "path": "command://myCommand",
  "requireCalls": Array [
    Array [
      "/home/whoiam/projects/my-whook-project/src/commands/myCommand",
    ],
    Array [
      "/var/lib/node/node_modules/@whook/whook/src/commands/myCommand",
    ],
  ],
  "result": undefined,
}
`;

exports[`$autoload should work with project commands 1`] = `
Object {
  "baseAutoloaderCalls": Array [],
  "logCalls": Array [
    Array [
      "debug",
      "🤖 - Wrapping the whook autoloader.",
    ],
    Array [
      "warning",
      "Command called!",
    ],
  ],
  "path": "command://myCommand",
  "requireCalls": Array [
    Array [
      "/home/whoiam/projects/my-whook-project/src/commands/myCommand",
    ],
  ],
  "result": undefined,
}
`;

exports[`$autoload should work with whook-cli commands 1`] = `
Object {
  "baseAutoloaderCalls": Array [],
  "logCalls": Array [
    Array [
      "debug",
      "🤖 - Wrapping the whook autoloader.",
    ],
    Array [
      "debug",
      "Command \\"myCommand\\" not found in: /home/whoiam/projects/my-whook-project/src/commands/myCommand",
    ],
    Array [
      "warning",
      "Command called!",
    ],
  ],
  "path": "command://myCommand",
  "requireCalls": Array [
    Array [
      "/home/whoiam/projects/my-whook-project/src/commands/myCommand",
    ],
    Array [
      "/var/lib/node/node_modules/@whook/whook/src/commands/myCommand",
    ],
  ],
  "result": undefined,
}
`;
